src/tsdb.c

```diff
@@ -224,85 +225,128 @@ int SeriesAddSample(Series *series, api_timestamp_t timestamp, double value) {
+            if (funcs->GetFirstTimestamp(iter->currentChunk) > iter->maxTimestamp) {
+                dictResult = RedisModule_DictPrevC(iter->dictIter, NULL, (void*)&iter->currentChunk);
+            } else break;
+        }      
     }
-    SeriesIteratorClose(iter);
-    return REDISMODULE_ERR;
+    
+    if (dictResult == NULL) {
+        return REDISMODULE_ERR;
+    }
+    iter->chunkIterator = funcs->NewChunkIterator(iter->currentChunk, iter->reverse);
+    return REDISMODULE_OK;
 }
 
+// Fill the first sample
 ChunkResult SeriesIteratorGetFirst(SeriesIterator *iterator, Sample *sample) {
+    ChunkResult res;
     ChunkFuncs *funcs = iterator->series->funcs;
-    ChunkResult res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
 
     if (funcs->GetNumOfSample(iterator->currentChunk) == 0) {
         return CR_END;
     }
 
-    if (sample->timestamp < iterator->minTimestamp) {
-        while (res == CR_OK) {     
-            // Skip through initial samples
-            res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
-            if (sample->timestamp >= iterator->minTimestamp) { 
-                break;
+    if (iterator->reverse == NO_OPT) {
+        res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
+        if (sample->timestamp < iterator->minTimestamp) {
+            while (res == CR_OK) {     
+                // Skip through initial samples
+                res = funcs->ChunkIteratorGetNext(iterator->chunkIterator, sample);
+                if (sample->timestamp >= iterator->minTimestamp) { 
+                    break;
+                }
+            }
+            return res; // shouldn't happen
+        }
+    } else {
+        res = funcs->ChunkIteratorGetPrev(iterator->chunkIterator, sample);
+        if (sample->timestamp > iterator->maxTimestamp) {
+            while (res == CR_OK) {     
+                // Skip through initial samples
+                res = funcs->ChunkIteratorGetPrev(iterator->chunkIterator, sample);
+                if (sample->timestamp <= iterator->maxTimestamp) { 
+                    break;
+                }
             }
+            return res; // shouldn't happen
         }
-        return res; // shouldn't happen
     }
     // No sample within range
-    if (sample->timestamp > iterator->maxTimestamp) {
+    if (sample->timestamp > iterator->maxTimestamp ||
+        sample->timestamp < iterator->minTimestamp) {
         return CR_END;   
     } 
     return CR_OK;
 }
 
 void SeriesIteratorClose(SeriesIterator *iterator) {
-    if (iterator->chunkIterator != NULL)
-        iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator);
-    if (iterator->dictIter != NULL)
-        RedisModule_DictIteratorStop(iterator->dictIter);
-    *iterator = (SeriesIterator){ 0 };
+    iterator->series->funcs->FreeChunkIterator(iterator->chunkIterator, iterator->reverse);
+    RedisModule_DictIteratorStop(iterator->dictIter);
```

Comment:
please reintroduce the iterator zeroing, it will allow us to spot bugs where we reuse iterators and have garbage in them:
```
    *iterator = (SeriesIterator){ 0 };	
```

Reply:
@danni-m 
Reintroduce? Where is it like this??